### ISSUE TITLE
abi_l2_nc reader unable to read MCMIP files

### ISSUE THREAD

**Issue by paapu88** [pytroll/satpy#1162]
Dear Developers,
I'm using satpy 0.21 from this git-repo.
But I cannot read with satpy abi_l2_nc reader GOES16 data.
So can one read GOES16 ABI-L2 data with satpy, have I just not installed the reader?
Cheers, Markus
below the data is read fine with xarray but I have problems wih satpy...
Can one somehow set the data read by xarray into satpy into 

```
 self.data = xarray.open_dataset(fileName)
 print("DATA KEYS:", self.data.keys())
 scn = Scene(reader='abi_l2_nc', filenames=[fileName])
```
xarray reads the data fine but satpy gives an error.

> DOWNLOAD: noaa-goes16/ABI-L2-MCMIPF/2019/244/18/OR_ABI-L2-MCMIPF-M6_G16_s20192441800192_e20192441809505_c20192441810011.nc
File already there (not downloading): /home/hu-mka/Downloads/OR_ABI-L2-MCMIPF-M6_G16_s20192441800192_e20192441809505_c20192441810011.nc
DATA KEYS: KeysView(<xarray.Dataset>
Dimensions:                                 (band: 1, number_of_image_bounds: 2, number_of_time_bounds: 2, x: 5424, y: 5424)
Coordinates:
    t                                       datetime64[ns] ...
    y                                       (y) float32 0.151844 ... -0.151844
    x                                       (x) float32 -0.151844 ... 0.151844
    y_image                                 float32 ...
    x_image                                 float32 ...
    band_wavelength_C01                     (band) float32 ...
    band_wavelength_C02                     (band) float32 ...
    band_wavelength_C03                     (band) float32 ...
    band_wavelength_C04                     (band) float32 ...
    band_wavelength_C05                     (band) float32 ...
    band_wavelength_C06                     (band) float32 ...
    band_wavelength_C07                     (band) float32 ...

> No filenames found for reader: abi_l2_nc
Don't know how to open the following files: {'/home/hu-mka/Downloads/OR_ABI-L2-MCMIPF-M6_G16_s20192441800192_e20192441809505_c20192441810011.nc'}

**Comment by djhoese** [comment]
Your filenames have an "M" in front of "CMIPF", any idea what that means? The reader in Satpy is not configured to handle these files and I've never seen them before. If you have more information about them then we can definitely modify the reader to support them.

FYI the current file patterns that the reader knows about: https://github.com/pytroll/satpy/blob/master/satpy/etc/readers/abi_l2_nc.yaml#L325

**Comment by paapu88** [comment]
Thanks for a quick reply and sorry for my wrong label. 
More info is in https://docs.opendata.aws/noaa-goes16/cics-readme.html (the product ABI-L2-MCMIPF). 

Maybe I can use a product where there is a reader available. Product is copy pasted from https://github.com/blaylockbk/pyBKB_v3/blob/master/BB_GOES/GOES16_TrueColor_Cartopy.ipynb

**Comment by paapu88** [comment]
It seems that product 'ABI-L2-CMIPF' for which there is a reader has only one channel:
>Dimensions without coordinates: band, number_of_image_bounds, number_of_time_bounds
Data variables:
    CMI                                               (y, x) float32 ...
    DQF                                               (y, x) float32 ...

But I'd like to get all the 16 channels available (product 'ABI-L2-MCMIPF')
> Data variables:
    CMI_C01                                 (y, x) float32 ...
    DQF_C01                                 (y, x) float32 ...
    CMI_C02                                 (y, x) float32 ...
    DQF_C02                                 (y, x) float32 ...
    CMI_C03                                 (y, x) float32 ...
   .....

So should one write a new reader or is there a work-around?

**Comment by djhoese** [comment]
Right, the CMIP files are 1 channel per file, 16 files per time step. I didn't know the format you are using existed but I think there is a simple solution. If you're comfortable making a pull request it would be great to add official support for these files in the `abi_l2` reader. I can help.

1. Make a new `file_type` for your file pattern. Something like:

   ```
   abi_l2_mcmip:
        file_reader: !!python/name:satpy.readers.abi_l2_nc.NC_ABI_L2
        file_patterns: ['{system_environment:2s}_{mission_id:3s}-L2-MCMIP{scene_abbr:s}-{scan_mode:2s}_{platform_shortname:3s}_s{start_time:%Y%j%H%M%S%f}_e{end_time:%Y%j%H%M%S%f}_c{creation_time:%Y%j%H%M%S%f}.nc']
   ```

2. Edit each of the `C01` through `C16` dataset sections and change file type to say:

   ```
   file_type:
       abi_l2_cmip_c01:
           file_key: CMI
       abi_l2_mcmip:
           file_key: CMI_C01
   ```

I think doing these changes should work assuming things are coding flexibly in the reader's python code.

**Comment by paapu88** [comment]
Sorry, it turned out that I cannot work at the moment for GPL license projects...

**Comment by duncanwp** [comment]
Hi @djhoese , I wanted to pick this up as I'm now using the same MCMIP data. I tried your suggestion above but I ran in to the folowing error:
```
  File "./process_goes_l2_se_pacific_hackathon.py", line 118, in load_and_save_vars
    scn.load(variables)
  File "/home/ubuntu/anaconda3/envs/satpy/lib/python3.7/site-packages/satpy/scene.py", line 973, in load
    self.read(**kwargs)
  File "/home/ubuntu/anaconda3/envs/satpy/lib/python3.7/site-packages/satpy/scene.py", line 880, in read
    return self._read_datasets(nodes, **kwargs)
  File "/home/ubuntu/anaconda3/envs/satpy/lib/python3.7/site-packages/satpy/scene.py", line 721, in _read_datasets
    new_datasets = reader_instance.load(ds_ids, **kwargs)
  File "/home/ubuntu/anaconda3/envs/satpy/lib/python3.7/site-packages/satpy/readers/yaml_reader.py", line 872, in load
    ds = self._load_dataset_with_area(dsid, coords, **kwargs)
  File "/home/ubuntu/anaconda3/envs/satpy/lib/python3.7/site-packages/satpy/readers/yaml_reader.py", line 775, in _load_dataset_with_area
    file_handlers = self._get_file_handlers(dsid)
  File "/home/ubuntu/anaconda3/envs/satpy/lib/python3.7/site-packages/satpy/readers/yaml_reader.py", line 721, in _get_file_handlers
    filetype = self._preferred_filetype(ds_info['file_type'])
  File "/home/ubuntu/anaconda3/envs/satpy/lib/python3.7/site-packages/satpy/readers/yaml_reader.py", line 684, in _preferred_filetype
    if filetype in self.file_handlers:
TypeError: unhashable type: 'dict'
```
Any ideas?

Happy to turn this in to a pull-request once I have it working.

**Comment by djhoese** [comment]
Hi @duncanwp, could you show me a diff of what you changed in the YAML? It may actually be best to start a PR so I can see everything all at once. Also note that @mraspaud merged a huge PR this morning (#1088) that changes a lot of the internals of Satpy and effects readers. You may want to re-apply your changes on top of current master.
